Android/Tapp/app/src/main/java/app/anish/com/tapp/fragments/LinkedinDialogFragment.java

```diff
@@ -76,144 +72,94 @@ private void initUIElements(View rootView) {
     }
 
-    private AccessToken retrieveAccessToken() {
-        String tokenString = SharedPrefsUtils.getString(context, AppConstants.SETTINGS_SHARED_PREFS_KEY, Token.LINKEDIN.toString());
-        if (tokenString != null) {
-            return AccessToken.buildAccessToken(tokenString);
-        }
-        return null;
-    }
-
-    private void toggleProgressBar() {
-        if (progressBar.getVisibility() == View.GONE) {
+    private void toggleProgressBar(boolean show) {
+        if (show) {
             progressBar.setVisibility(View.VISIBLE);
         } else {
             progressBar.setVisibility(View.GONE);
         }
     }
 
-    private void saveLinkedinInfo(JSONObject jsonObject) {
-        try {
-            String linkedinId = jsonObject.getString("id");
-            String firstName = jsonObject.getString("firstName");
-            String lastName = jsonObject.getString("lastName");
-            SharedPrefsUtils.saveString(context, AppConstants.SETTINGS_SHARED_PREFS_KEY, SecuredSharedPrefs.LINKEDIN_ID.getInfoPrefKey(), linkedinId);
-            SharedPrefsUtils.saveString(context, AppConstants.SETTINGS_SHARED_PREFS_KEY, SettingsInfo.LINKEDIN_NAME.getInfoPrefKey(),
-                    firstName + " " + lastName);
-
-        } catch (JSONException e) {
-            Toast.makeText(context, "Error retrieving Linkedin Info from result", Toast.LENGTH_LONG);
+    private void toggleLoginButton(boolean loggedIn) {
+        if (loggedIn) {
+            loginButton.setText(R.string.linkedin_logout_text);
+            loginButton.setTag(R.integer.linkedin_login_status, true);
+        } else {
+            loginButton.setText(R.string.linkedin_login_text);
+            loginButton.setTag(R.integer.linkedin_login_status, false);
         }
     }
 
-    private void logoutOfLinkedin() {
+    private void logoutOfLinkedIn() {
         // remove the linkedin id and access token from shared preferences
         SharedPrefsUtils.deleteKey(context, AppConstants.SETTINGS_SHARED_PREFS_KEY, Token.LINKEDIN.toString());
         SharedPrefsUtils.deleteKey(context, AppConstants.SETTINGS_SHARED_PREFS_KEY, SecuredSharedPrefs.LINKEDIN_ID.getInfoPrefKey());
         SharedPrefsUtils.deleteKey(context, AppConstants.SETTINGS_SHARED_PREFS_KEY, SettingsInfo.LINKEDIN_NAME.getInfoPrefKey());
+        toggleLoginButton(false);
+    }
+
+    @Override
+    public void onClick(View view) {
+        switch (view.getId()) {
+            case R.id.bLinkedinButton:
+                processLoginButtonClick();
+                break;
+        }
+    }
+
+    private class MyFlowCompletionListener implements FlowCompletionListener {
+
+        @Override
+        public void onSuccess() {
+            toggleLoginButton(true);
+            toggleProgressBar(false);
+        }
+
+        @Override
+        public void onFailure(Throwable throwable) {
+            Toast.makeText(context, "Failed to login into Linkedin", Toast.LENGTH_SHORT).show();
```

Comment:
Yes. The simplest example is changing the orientation of the screen. For example, if the screen orientation is in portrait and.
1. The async task is triggered
2. The screen is rotated
3. The async task completes

The context that will be used when the AsyncTask completes will be the context that was valid prior to rotating the screen. When the screen rotates, the Android system takes the activity lifecycle all the way through destroying the activity. Then it recreates the activity, recalculating the size and positioning of the views on the screen for the new orientation.

Thus, the context that you would really want the Android system to use is the new context after rotation--not the old context.

And no. You don't always want to use ```getApplicationContext()```. You want to use the correct context for the operation you're performing. We can get together and talk about this.